### ISSUE TITLE
otelgrpc high metric cardinality from rpc.server.duration

### ISSUE THREAD

**Issue by rcrowe** [open-telemetry/opentelemetry-go-contrib#3071]
Since upgrading to **v1.12.0/0.37.0/0.6.0** & specifically https://github.com/open-telemetry/opentelemetry-go-contrib/pull/2700 we've seen an explosion in the cardinality of our metrics.

1. Metric recorded: https://github.com/open-telemetry/opentelemetry-go-contrib/blob/ad04d9c71cb8974374ba8e3ee454fdaaddc3e345/instrumentation/google.golang.org/grpc/otelgrpc/interceptor.go#L346
2. Attributes are taken from `spanInfo`: https://github.com/open-telemetry/opentelemetry-go-contrib/blob/ad04d9c71cb8974374ba8e3ee454fdaaddc3e345/instrumentation/google.golang.org/grpc/otelgrpc/interceptor.go#L331
3. Peer attributes added: https://github.com/open-telemetry/opentelemetry-go-contrib/blob/ad04d9c71cb8974374ba8e3ee454fdaaddc3e345/instrumentation/google.golang.org/grpc/otelgrpc/interceptor.go#L476-L479

Is the intention that these types of high cardinality issues are handled in userland, by us using a view (or blocking at the ingestion layer) to strip labels from certain metrics after we find them problematic? On one hand I can understand that being acceptable, on the other, this feels like pushing a problem downstream to consumers that will suddenly get hit by a metric that is most likely going to grow very wide.

CC @fatsheep9146

**Comment by rcrowe** [comment]
Missed https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/semantic_conventions/rpc-metrics.md#attributes which seems to suggest this is working as expected

**Comment by lucasoares** [comment]
How to remove this metric? It is exposing 7000 metrics in the exporter.

**Comment by regeda** [comment]
After `v0.38.0` release, I see the same issue with `otelhttp` package. We avoid this upgrade bc it blows up the memory consumption drastically.

**Comment by MrAlias** [comment]
You can use a [`View`](https://pkg.go.dev/go.opentelemetry.io/otel/sdk/metric#NewView) to drop the aggregation entirely or filter the attributes causing your cardinality explosion.

**Comment by regeda** [comment]
> You can use a [`View`](https://pkg.go.dev/go.opentelemetry.io/otel/sdk/metric#NewView) to drop the aggregation entirely or filter the attributes causing your cardinality explosion.

Could you share an example how to drop certain attributes from the aggregation?

**Comment by MrAlias** [comment]
> > You can use a [`View`](https://pkg.go.dev/go.opentelemetry.io/otel/sdk/metric#NewView) to drop the aggregation entirely or filter the attributes causing your cardinality explosion.
> 
> Could you share an example how to drop certain attributes from the aggregation?

The linked documentation has an example that should help as a basis:

![20230203_085416](https://user-images.githubusercontent.com/5543599/216661257-7d416621-677f-4a6c-9b55-113f33a8d1fa.png)

You'll need to update your match criteria to specifically match what you want though.

**Comment by regeda** [comment]
I've done with the following example (thanks @MrAlias ):
```go

func allowedAttr(v ...string) attribute.Filter {
	m := make(map[string]struct{}, len(v))
	for _, s := range v {
		m[s] = struct{}{}
	}
	return func(kv attribute.KeyValue) bool {
		_, ok := m[string(kv.Key)]
		return ok
	}
}

func setGlobalMeterProvider() error {
	exporter, err := prometheus.New(
		prometheus.WithRegisterer(client_prom.DefaultRegisterer),
	)
	if err != nil {
		return err
	}
	global.SetMeterProvider(metric.NewMeterProvider(
		metric.WithReader(exporter),
		metric.WithView(
			metric.NewView(
				metric.Instrument{Scope: instrumentation.Scope{Name: "go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp"}},
				metric.Stream{
					AttributeFilter: allowedAttr(
						"http.method",
						"http.status_code",
						"http.target",
					),
				},
			),
		),
	))
	return nil
}

```

**Comment by regeda** [comment]
Another issue :(

A new version of `otelhttp/v0.38.0` uses `go.opentelemetry.io/otel/semconv/v1.17.0/httpconv.ServerRequest` for attributes generating https://github.com/open-telemetry/opentelemetry-go-contrib/blob/main/instrumentation/net/http/otelhttp/handler.go#L218.

This function writes the whole request URI into `http.target`. It impacts the server's memory if the URI is constantly random.

**Comment by regeda** [comment]
Also, the view's override doesn't help too much bc the internal aggregation is still happening with the following attributes that impact memory consumption and there is no option to prevent it:
- peer port
- user agent
- request uri

**Comment by MrAlias** [comment]
> Also, the view's override doesn't help too much bc the internal aggregation is still happening with the following attributes that impact memory consumption and there is no option to prevent it:
> 
> * peer port
> * user agent
> * request uri

Can you expand on this? I'm not sure how an attribute filter wouldn't help prevent this. The filter wraps the aggregation so attributes filtered out should never reach the underlying aggregation storage.

**Comment by lucasoares** [comment]
I just want to be able to disable 100% of grpc and http metrics from ever being created, to not have any kind of impact on my application.

Any way to do it? I don't want to have to do bench tests because of these third-party metrics...

**Comment by MrAlias** [comment]
> I just want to be able to disable 100% of grpc and http metrics from ever being created, to not have any kind of impact on my application.
> 
> Any way to do it? I don't want to have to do bench tests because of these third-party metrics...

https://github.com/open-telemetry/opentelemetry-go-contrib/issues/3071#issuecomment-1416137206

**Comment by liufuyang** [comment]
@MrAlias I think I encounter the same memory issue. So have this https://github.com/open-telemetry/opentelemetry-go/pull/3695 fixed it and does the current release includes it? Thank you :)

**Comment by jaronoff97** [comment]
Hello all, we're running a collector right now and we're seeing this issue cause a massive memory leak when converting the metric to prometheus. This stems from the new usage of the meter provider [here](https://github.com/open-telemetry/opentelemetry-collector/blob/a2f0153679220d5de7b2a2ac79a5bebfca90cf89/config/configgrpc/configgrpc.go#L263) in the collector. 

IMO this is a massive issue because it means someone could cause OOMs/dropped data by flooding the collector with bad rpcs. Would it be possible to make the inclusion of the `net.sock.peer.addr`, `net.sock.peer.port`, etc. optional? The inclusion of these attributes are useful for traces, but cause massive cardinality explosion in metrics. Because the collector also is exporting Prometheus Histograms, the cardinality we observe is `|net.sock.peer.addr|` x `|net.sock.peer.port|` x `|buckets|` which results in millions of metrics created.

Users of the collector are unable to implement a custom view as has been recommended on this thread, because the code is already built with this. Given other users are experiencing this issue, it felt prudent to upstream this request to the source.

**Comment by mackjmr** [comment]
I experienced this when running `opentelemetry-demo`. 

Here is a graph showing the memory usage of the collector pod before / after implementing the filter from https://github.com/open-telemetry/opentelemetry-go-contrib/issues/3071#issuecomment-1419366500 in [productcatalogservice](https://github.com/open-telemetry/opentelemetry-demo/tree/main/src/productcatalogservice) (fix pushed at 17h25): 

<img src="https://github.com/open-telemetry/opentelemetry-go-contrib/assets/63265430/c8e36581-c550-4e7f-87ce-5a10d5f3eecc" width= 482 height=401/>

IMO, `net.sock.peer.port` should be disabled by default seeing the cardinality / significant memory usage in the collector it leads to when enabled. 

cc @pellared as it seems like you are working on a fix to this issue in #3730.
